kong/clustering/services/sync/rpc.lua

```diff
@@ -413,12 +402,27 @@ function sync_once_impl(premature, retry_count)
 
   -- retry if the version is not updated
   retry_count = retry_count or 0
+
   if retry_count > MAX_RETRY then
     ngx_log(ngx_ERR, "sync_once retry count exceeded. retry_count: ", retry_count)
     return
   end
 
-  return start_sync_once_timer(retry_count + 1)
+  -- we do not count a timed out sync. just retry
+  if err ~= "timeout" then
+    retry_count = retry_count + 1
+  end
+
+  -- in some cases, the new spawned timer will be switched to immediately,
+  -- preventing the coroutine who possesses the mutex to run
+  -- to let other coroutines has a chance to run
+  local ok, err = kong.timer:at(0.1, sync_once_impl, retry_count or 0)
+  -- this is a workaround for a timerng bug, where tail recursion causes failure
```

Comment:
Is it a bug? @ADD-SP could you confirm it?

Reply:
@chronolaw I've talked to @ADD-SP offline, and confirmed it's a bug: in tail recursion, the stack info cannot be retrieved and it causes timer creation to crash as it assume at least 1 frame to exist

Reply:
Yes, tail call optimization eliminates a call stack frame, so we might get an empty call stack if we create a timer at the entry function that has no caller (like called by C code).